R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
-# Simple interface for training an xgboost model that wraps \code{xgb.train}.
-# Its documentation is combined with xgb.train.
-#
-#' @rdname xgb.train
+prescreen.parameters <- function(params) {
+  if (!NROW(params)) {
+    return(list())
+  }
+  if (!is.list(params)) {
+    stop("'params' must be a list or NULL.")
+  }
+
+  params <- params[!is.null(params)]
+
+  if ("num_class" %in% names(params)) {
+    stop("'num_class' cannot be manually specified for 'xgboost()'. Pass a factor 'y' instead.")
+  }
+  if ("process_type" %in% names(params)) {
+    if (params$process_type != "default") {
+      stop("Non-default 'process_type' is not supported for 'xgboost()'. Try 'xgb.train()'.")
+    }
+  }
+
+  params_function_args <- c(
+    "objective", "verbose", "verbosity", "nthread", "seed",
+    "monotone_constraints", "interaction_constraints"
```

Comment:
They are meant to be passed as arguments to the function `xgboost()` instead of being passed through `params`.

Reply:
`monotone_constraints` and `interaction_constraints` might be better as a part of normal params? This way it feels more consistent. It's difficult from a user's perspective to understand why some of these parameters are different.

Reply:
From my perspective:
* Some parameters I chose to put as function arguments in order to give them a default value visible to the user (e.g. `nthreads`) or make the default value change according to the data (e.g. `objective`).
* Some parameters are independent of the data (e.g. regularization) or work regardless of whether the data supports them or not (e.g. max categ. to one-hot), while others refer specifically to columns in the data (e.g. interaction constraints) and would not work with a different dataset.
* Ideally, later on there should be a function like `xgb.control` or similar that would list the parameters and offer auto-completion by having them as function arguments (e.g. like `VGAM::vglm.control` or `glmx::glmx.control`). Those train-control parameters typically involve things that are independent of the data.
* Alternatively, and as is more common in R, _all_ parameters could be moved to function arguments (e.g. like `ranger::ranger` or `glmnet::glmnet`), but perhaps that would make the docs very hard to use.

> It's difficult from a user's perspective to understand why some of these parameters are different.

Yes, it's a bit arbitrary, but that is also the situation in the other interfaces - e.g. the number of boosting rounds is a function argument, as is `verbose_eval` in the python interface. It also feels like parameters such as `seed` should be in the train control list, but that would have other side effects due to the logic introduced in an earlier PR (https://github.com/dmlc/xgboost/pull/10029).

Reply:
Perhaps for v3 you could consider having a division between data-dependent parameters (e.g. `feature_weights` is passed under `DMatrix`, while `monotone_contraints` is passed under `params`) and data-independent parameters.

Reply:
Thank you for the detailed explanation. Please think of me as a newcomer instead of a maintainer. Or think of the future GitHub issues we will need to reply to when someone puts the parameter in the wrong place and asks why.

>  but that is also the situation in the other interfaces

I think that's only true for the low-level interface, and some of them are only in this way because we think of it as an unnecessary breaking change. For higher-level interfaces like the scikit-learn, or Spark, they have a relatively consistent interface. For the scikit-learn, we are moving the parameters according to the scikit-learn estimator guideline where the constructor takes hyper-parameter and sometimes feature-related data like feature types, while the `fit` method takes data(sample)-dependent parameters that can be split for early stopping, etc.

> default value visible to the user

We can compromise on this, maybe the document is sufficient?

> Alternatively, and as is more common in R, all parameters could be moved to function arguments

That sounds reasonable to me. Personally, I prefer consistency, which has been a pain point for XGBoost for a while now. Would be great to see improvement in this area.

> Perhaps for v3 you could consider having a division between data-dependent parameters (e

Hmm, since this is a new interface already, maybe it's better to do it right the first time if we know where the issue is?

Reply:
Ok then, I'll move all parameters to function arguments, leaving them as undocumented `...` for now (to be changed in a later PR). We'll see how the docs and examples end up looking later on.

Reply:
Modified it by removing `params` and making all parameters as function arguments, by using `...` for now.